viewing_party/party.py

```diff
@@ -1,23 +1,225 @@
 # ------------- WAVE 1 --------------------
+import pprint
 
 def create_movie(title, genre, rating):
-    pass
+    """Create dictionary containing title, genre, and rating of a movie
+    
+    Keyword arguments:
+    title -- the title of the movie
+    genre -- the genre of the movie
+    rating -- the rating of the movie
+    """
+    # Return None if any parameter is falsy
+    if not title or not genre or not rating:
+        return None
+        
+    movie = {
+        "title": title, 
+        "genre": genre,
+        "rating": rating
+    }
+
+    return movie
+
+def add_to_watched(user_data, movie):
+    """Add a movie to watched list.
+    
+    Keyword arguments:
+    user_data -- dictionary containing movies user has and wants to watch
+    movie -- dictionary containing title, genre, and rating
+    """
+    # Check movie isn't empty and is a dictionary
+    # Try to add movie to watched, otherwise give error message
+    if movie and isinstance(movie, dict):
+        try:
+            user_data["watched"].append(movie)
+        except KeyError:
+            print("'Watched' key does not exist; could not add movie")
+    
+    return user_data
+
+def add_to_watchlist(user_data, movie):
+    """Add a movie to a watchlist.
+    
+    Keyword arguments:
+    user_data -- dictionary containing movies user has and wants to watch
+    movie -- dictionary containing title, genre, and rating
+    """
+    # Check movie isn't empty and is a dictionary
+    # Try to add movie to watchlist, otherwise give error message
+    if movie and isinstance(movie, dict):
+        try:
+            user_data["watchlist"].append(movie)
+        except KeyError:
+            print("'Watchlist' key does not exist; could not add movie")
+    
+    return user_data
+
+def watch_movie(user_data, title):
+    """Move movie if in watchlist to watched
+    
+    Keyword arguments:
+    user_data -- dictionary containing movies user has and wants to watch
+    title -- the movie title 
+    """
+    # Check title is a string and not empty, otherwise return user_data
+    if not isinstance(title, str) or not title:
+        return user_data
+
+    # Check if movie with title in watchlist
+    # If it is, append movie to watched and remove from watchlist
+    for movie in user_data["watchlist"]:
+        if title in movie["title"]:
+            user_data["watched"].append(movie)
+            user_data["watchlist"].remove(movie)

```

Comment:
Here we are iterating over the user's watchlist, and also modifying it (`remove`). We should avoid modifying a list while iterating over it to avoid problems with the iteration. _If_ we stop iterating as soon as we make a change, it would still be safe (though anyone else reading the code would need to convince themselves of this). However, depending on our data restrictions, this code could end up modifying the list multiple times if there is more than one movie with the same title.

To ensure that only one change is made, we could break out as soon as we find the movie and move it. Again, this is safe from an iteration perspective, but may require other team members to spend time convincing themselves of this. We might instead decide to think of this as 2 steps. 1. Find the movie to move by iterating through the list, and 2. actually move the move (outside the list). This might be more clear that there are no negative modifying while iterating shenanigans!

```python
        for movie in user_data["watchlist"]:
            if title in movie["title"]:
                user_data["watched"].append(movie)
                user_data["watchlist"].remove(movie)
                break

        return user_data
```